src/main/java/com/couchbase/lite/Database.java

```diff
@@ -1152,27 +1149,54 @@ private void markBlobstoreMigrated() {
      */
     @InterfaceAudience.Private
     public boolean close() {
-        if(!open) {
+        if (!open) {
             return false;
         }
 
-        if(views != null) {
+        if (views != null) {
             for (View view : views.values()) {
                 view.databaseClosing();
             }
         }
         views = null;
 
-        if(activeReplicators != null) {
-            for(Replication replicator : activeReplicators) {
+        if (activeReplicators != null) {
+            List<CountDownLatch> latches = new ArrayList<CountDownLatch>();
+            for (Replication replicator : activeReplicators) {
+                // handler to check if the replicator stopped
+                final CountDownLatch latch = new CountDownLatch(1);
+                replicator.addChangeListener(new Replication.ChangeListener() {
+                    @Override
+                    public void changed(Replication.ChangeEvent event) {
+                        if (event.getSource().getStatus() == Replication.ReplicationStatus.REPLICATION_STOPPED) {
+                            latch.countDown();
+                        }
+                    }
+                });
+                latches.add(latch);
+
+                // ask replicator to stop
                 replicator.databaseClosing();
             }
+
```

Comment:
For #1, what I mean is that if activeReplicators property is not thread-safe, there could be a case that while setting up the change listener for a replicator, that replicator might be stopped due to some reasons (e.g. permanent error or finish replicating (non-continuous)) -- actually if that happens, there will be a runtime exception thrown for iterating over a changed array.